packages/components/src/navigator/match-path.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * External dependencies
+ */
+import pathToRegexp from 'path-to-regexp';
+
+const cache = {};
+const cacheLimit = 10000;
+let cacheCount = 0;
+
+function compilePath( path, options ) {
+	const cacheKey = `${ options.end }${ options.strict }${ options.sensitive }`;
+	const pathCache = cache[ cacheKey ] || ( cache[ cacheKey ] = {} );
+
+	if ( pathCache[ path ] ) return pathCache[ path ];
+
+	const keys = [];
+	const regexp = pathToRegexp( path, keys, options );
+	const result = { keys, regexp };
+
+	if ( cacheCount < cacheLimit ) {
+		pathCache[ path ] = result;
+		cacheCount++;
+	}
+
+	return result;
+}
+
+/**
+ * Public API for matching a URL pathname to a path.
+ *
+ * @param {string}       pathname Pathname
+ * @param {string|Array} options  Matching options.
```

Comment:
I guess here we should document we accept {string|Array|Object}. And the documentation should say something like the path to be matched against pathName or an object specifying the path to be matched and the matching options.